File: backend/model/credit.ts
import {v4 as uuid} from 'uuid'
import Transaction from "./transaction";
import Account from "./account"

export default class Credit extends Transaction {
  static Of(amount: number): Credit {
    return new Credit(uuid(), amount, new Date())
  }
  
  type(): string {
    return 'credit'
  }

  applyTo(account: Account): void {
    account.deposit(this._amount)
  }
}

File: backend-golang/model/credit.go
package model

import (
	"github.com/google/uuid"
	"time"
)

const TRANSACTION_TYPE_CREDIT = "credit"

type Credit Transaction

func NewCredit(amount float32) *Credit {
	return &Credit{
		Id:            uuid.New().String(),
		Type:          TRANSACTION_TYPE_CREDIT,
		Amount:        amount,
		EffectiveDate: time.Now(),
	}
}

func (credit Credit) ApplyTo(account *Account) error {
	account.Balance += credit.Amount
	credit.EffectiveDate = time.Now()
	return nil
}
